README.md

```diff
@@ -149,29 +144,30 @@ This requires:
    be possible to use compressed sets with only partial hashes, and a confirmation API in case of a hit.
+ * An interesting way to store geographic (location) data may be to use Rabin-Karp fingerprinting based on a GeoHash, probably more relevant to the front-end, not back-end.
 
 # Very tentative design ideas
 ## Submission
 Reporting a infection is a simple POST to a very redundant server that logs
 the post as durably as possible. A POST might not be accepted because it has
 no valid matching healthcare provider token, so we can't just log POSTs and
-hope for the best. Some interaction is required it appears. 
+hope for the best. Some interaction is required it appears. --> COMMENT: Yes, it's still client/server for reporting, so clients also need to handle proper error handling and submission retry mechanisms. Logging is needed so we can monitor the rate of succeeding and failing back-end transmissions (and intervene and recover). It may be advisable to run multiple API deployments that use another mechanism to synchronize their back-end submissions to avoid a single point of failure.
 
 If we could be somewhat asynchronous we could log POSTS without interaction
 and have a 100% up collector. Very tempting. This would require the app to
 ask back to see if the report was accepted. Might be worth it. We could
-provide a tracking ID in response to a POST.
+provide a tracking ID in response to a POST --> COMMENT: Yes, that may work well with the multiple endpoints/syncing ID above.
 
 ## Polling
 It would be SUPER ACE if we could put the data files on a CDN and not have
 to worry about them anymore. The initial data file would be quite large, and
-you'd have to apply delta-sets to them.
+you'd have to apply delta-sets to them -> COMMENT: As suggested, the back-end could write out it's stuff to a bucket in Delta Lake format, which clients can read from. This bucket could be replicated. This is probably the easiest mechanism for distribution to clients (read: apps) I can think of.
+
+There is no need to do realtime transmission of updates, it is acceptable to have some delay. Perhaps delta files per 15 minutes? And always also generate a full dump so 'dumb apps' could use that instead? --> COMMENT: I'd go the dumb route only when necessary, the more there's one 'ticket booth' so to speak, the better that generally is.
+
+File need built-in integrity check so we know we got a full file correctly --> COMMENT: Delta Lake, otherwise: CRC/Adler/MD5/MurmurHash (any non-cryptographic hash) or something like that over the data files.
 
-There is no need to do realtime transmission of updates, it is acceptable to
-have some delay. Perhaps delta files per 15 minutes? And always also
-generate a full dump so 'dumb apps' could use that instead?
+Note that delta-sets also include deletion events, either because the key is no longer relevant (post 14 days) or because we corrected a mistake --> COMMENT: Yes, this is a standard part of any transaction log :)
```

Comment:
> Why are the explicit deletions for expired keys needed in this case? I think the client app should be perfectly capable of determining the validity of a key and we should refrain from that on the backend

They are needed, because medical professionals might make mistakes in signing up clients.
It also lowers the bar for them if we can give them some kind of control over what they've put into our systems. This means they will be more likely to use it.

Reply:
So, there are two routes here (1) the validity is part of the key itself (or distributed along with it) and (2) there is some pre-set global rule that all keys expire after N days. 

The former (1) requires more data (one expiry per key), but is simple to handle, either by the clients (as you suggest) or by writing a transaction log entry to sort of enforce it (less complex client logic, enforces everyone has the same view at the same time). Something can be set for either approach.

The latter (2) allows us to change the data retention period overall (which can be advantageous if insights about how long the data needs to be retained changes - but it's also a bit of a privacy loophole) and is less complex to implement.

I am leaning towards option (1) with server side (transaction log) deletes for consistency reasons ... but perhaps I am missing something.

Reply:
I absolutely concur that a way of revocation is a must have. For testing and human mistakes. These are the exceptions if the normal process though. I think expiring records is a job for the app/client OR the healthcare provider. Our backend should not become part of a political debate on this point.
E.g:
1) backend has data for the last `X` (eg 21 days) <-- this is our domain, no extra work needs to be done, no extra data transmitted etc)
2) app builder (government/GGZ?) dicates a default of `Y` (eg 14 days). User of the app might be able to overrule this to be higher/lower? (here comes the politics:-))
3) The supplicant of the record (healthcare provider) can withdraw earlier for whatever reason they deem logical, using the revocation process

Side note: the Delta Files published should only contain the LAST record of a key. E.g a 'day' file with a report and a later withdrawal in it should only contain the withdrawal to get the client synced up to the desired state.

Reply:
What you are missing, is that we can and should have both!

A person will be either dead or not infectious anymore after 46 days of being actively infectious. So essentially this provides us with an upper limit of how long we might have to store our keys on the apps and servers.
(We can also set that number to 14 days, but the longest infectious period I've heard about with COVID-19 is 46 days, so it makes sense to keep that duration in mind and flexible when designing this back end).

Also note that, as a consequence of the specifications, the key-files from our servers will be mostly useless after 24 hours, because no keys derived from that day-key will be broadcast by the users devices via bluetooth anymore. And because there are egocentric privacy and technological (free up space) incentives for the users to delete the logs with whom they have been in contact after N-days. The users that do keep our back end key-lists are keeping data around that has become useless. The users that keep their contact-tracing logs are only in possession of their own contact-tracing logs which have also become useless. Even if they also retain a copy of our back end keylists.

At the same time, we need to account for the mistakes that will be made by humans, within that fixed period and we also have to show that "there is a way back" to medical professionals. Don't underestimate how hard it will be to get most of them on board due to the fact that doctor-patient confidentiality is considered "almost holy".

So we need to show them an active way to correct "mistakes" and a "passive" way in which the system automatically destructs the data.

Reply:
> What you are missing, is that we can and should have both!

We agree on that point from the beginning. The only nuance I'm looking for is on whether the backend should have a role in _explicitely (on our own initiative)_ revoking keys because they are expired due to a metric _we_ have configured. (so we should not explicitly tell the clients that we are about to delete a record, which is what we were about to do if I red the initial proposal correctly).

In my opinion the consumers and providers of our data should decide on which timerange of data is (still) valid and we have only one limit: the max age of the data (46 days worst case is what you're proposing?)